cpp/src_prims/stats/histogram.h

```diff
@@ -175,21 +175,92 @@ void smemBitsHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
   CUDA_CHECK(cudaGetLastError());
 }
 
+DI void clearHashTable(int2* ht, int hashSize) {
+  for (auto i = threadIdx.x; i < hashSize; i += blockDim.x) {
+    ht[i] = {-1, 0};
+  }
+  __syncthreads();
+}
+
+DI int findEntry(int2* ht, int hashSize, int binId) {
+  int idx = binId % hashSize;
+  int t;
+  int count = 0;
+  while ((t = atomicCAS(&(ht[idx].x), -1, binId)) != -1 && t != binId) {
+    ++count;
+    if (count >= hashSize) {
+      idx = -1;
+      break;
+    }
+    ++idx;
+    if (idx >= hashSize) {
+      idx = 0;
+    }
+  }
+  return idx;
+}
+
+DI void flushHashTable(int2* ht, int hashSize, int* bins) {
+  __syncthreads();
+  for (auto i = threadIdx.x; i < hashSize; i += blockDim.x) {
+    if (ht[i].x != -1 && ht[i].y > 0) {
+      atomicAdd(bins + ht[i].x, ht[i].y);
+    }
+    ht[i] = {-1, 0};
+  }
+  __syncthreads();
+}
+
 template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
 __global__ void smemHashHistKernel(int* bins, const DataT* data, IdxT n,
                                    IdxT nbins, BinnerOp binner, int hashSize) {
-  //extern __shared__ int2 smem[];
-  ///@todo!
+  extern __shared__ int2 ht[];
+  int* needFlush = (int*)&(ht[hashSize]);
+  if (threadIdx.x == 0) {
+    needFlush[0] = 0;
+  }
+  clearHashTable(ht, hashSize);
+  IdxT tid = threadIdx.x + IdxT(blockDim.x) * blockIdx.x;
+  IdxT stride = IdxT(blockDim.x) * gridDim.x;
+  int nCeil = ceildiv<int>(n, stride);
+  for (auto i = tid; i < nCeil; i += stride) {
+    bool iNeedFlush = false;
+    int binId = 0;
+    if (i < n) {
+      binId = binner(data[i], i);
+      int hidx = findEntry(ht, hashSize, binId);
+      if (hidx >= 0) {
+        atomicAdd(&(ht[hidx].y), 1);
+      } else {
+        needFlush[0] = 1;
+        iNeedFlush = true;
+      }
+    }
+    __syncthreads();
+    if (needFlush[0]) {
+      flushHashTable(ht, hashSize, bins);
+      if (threadIdx.x == 0) {
+        needFlush[0] = 0;
+      }
+      __syncthreads();
+    }
+    if (iNeedFlush) {
+      int hidx = findEntry(ht, hashSize, binId);
+      atomicAdd(&(ht[hidx].y), 1);
```

Comment:
Are you guaranteed to get a valid index here?  I think you can make a solid argument that after the flush, there are 6047 bins available, and that there should be enough room for every thread with iNeedFlush to get a bin.  I'll bet it is true 99.999999999% of the time.  Imagine a warp not getting scheduled for a while and the other warps fill the hashtable before this warp is scheduled.  Couldn't findEntry() return -1?

Reply:
Or maybe not.  If the loop goes around again, each thread in all the other warps will find a bin, increment it, and hit the syncthreads() call.  At that point, the thread with iNeedFlush set will have to make progress, so at most 1024-32 bins will have been used.  I think you're ok.

Reply:
Yup agreed. We don't need another check for valid index here. Thanks.

Reply:
I have left a comment at this place to help remind ourselves of this in the future.